### PR TITLE
thumbnails overhaul

### LINKED ISSUE #540: don't upscale images during thumbnail creation
There should be an option to not create new images when the thumbnail would generate an upscaled version, but return the original image instance instead.

I propose an `upscale=False` kwarg to `image.thumbnail()`, and plan to default it to `True` at some point in the future.

### PR DESCRIPTION
**Pull request by xlotlu**
- fix mismatch between reported thumbnail size and on-disk image when both width & height are provided;
- width can now be optional if height is provided;
- introducing thumbnail operation modes: "fit" (default. requested width / height behave as maximum allowed dimensions), "crop" (replaces previous `crop` kwarg) and "stretch" (aspect ratio isn't preserved);
- thumbnail returns original image when it's smaller than the requested dimensions and new `upscale` kwarg is `False` (default in "fit" mode).

other implementation details / smaller fixes:

- `get_image_info` returns jpeg dimensions swapped when exif rotation is in effect;
- fix bug when image type is unknown;
- fix off-by-1px rounding discrepancy between reported thumbnail size and imagemagick output.

closes #540.

@nixjdm: I'll follow up with a commit that gets all the backwards compatibility in place, once we settle on the final form. That way you'll be able to simply revert it for the magical breaking-change release.

@skorokithakis: this will affect your thumbnail-generator plugin. Do chime in if you have suggestions.

@runfalk: this also affects your work in #497.

### CODE REVIEW

**Comment by xlotlu** [comment]
@nixjdm once you read through the change, consider that I also have the following changes in mind:

- [ ] 1. add a `format` kwarg, that allows converting from one image format to another. Maybe also add a new mode "no-op", that's selected automatically if it's only a format change.

- [ ] 2. big change: write out the thumbnail name with full details, i.e. both width & height regardless of which one was provided, *and* with real values, as opposed to requested values, the way it is now. This would avoid useless duplicates (see the similar thumbnails test). It's not really a breaking change, and it's not a complicated change, but it implies regeneration of thumbnails for a whole many people.

- [x] 3. optimisation: do `find_imagemagick` only once. Ideally this should be done once, at startup, outside any `process_image` request. If too complicated, then at least cache its value. 

- [ ] 4. in the fluffy bla category: [seam-carving](https://upload.wikimedia.org/wikipedia/commons/b/bf/Broadway_tower_seam_carving_interactive.svg) "liquid rescale" mode.

Should these go into this PR? If they get the green light, it would definitely make things easier for that final revertible commit.

Also:
- [x] 0. include enum34 as dependency?

File: lektor/imagetools.py
Comment (nixjdm) [inline lektor/imagetools.py]: When does this fail? I can import this in 2.7 and 3.6. The failure isn't covered by the tests either.

**Review by nixjdm — CHANGES_REQUESTED** [review]
In addition to the other notes, I'd like you to add a few more test cases:

* That hit raised errors and warnings that are new. There's a few in `make_thumbnails` and one in `thumbnail`
* if source_width < width
* if source_height < height

Also about tests, it may be a good idea to create a dedicated demo project just for image tests, so that other build tests don't have to rebuild any of the images.

**Comment by nixjdm** [comment]
This is really good. I put some additional comments on a review.

About your second comment:

1. I like it.
2. I like this too.
3. You're on a roll.
4. Still good.

I agree, this would probably make the revertible commit easier, and I'm fine adding them to this. We can overhaul further.

0. I think I need a little explanation of this one. I'm missing something. When does the import fail?

Once this is all done, I also request help updating the docs. They'll certainly need it. :)

File: lektor/imagetools.py
Comment (xlotlu) [reply lektor/imagetools.py:478]: I know, it's just defensive programming. For the future and for plugins that might call it directly ([lektor-thumbnail-generator](https://github.com/skorokithakis/lektor-thumbnail-generator) does so for example).

File: lektor/imagetools.py
Comment (xlotlu) [reply lektor/imagetools.py:494]: No, in that example it's needed for shell escaping, which is not the case in python. The geometry syntax is http://www.imagemagick.org/script/command-line-processing.php#geometry -- when it's quoted the backslash would be an error.

File: lektor/imagetools.py
Comment (xlotlu) [reply lektor/imagetools.py:576]: And in... wait for it... the thumbnail-generator plugin. :)
I personally wouldn't rename it, because well, I'm too lazy to come up with something meaningful, and "compute_fit_dimensions" is perfectly meaningless. And anything else doesn't compute anything, it just applies them bluntly. Your choice.

**Comment by xlotlu** [comment]
@nixjdm, parse error, help me out here.
> * That hit raised errors and warnings that are new. There's a few in make_thumbnails and one in thumbnail

You mean to test the warnings and exceptions?

> * if source_width < width
> * if source_height < height

What of these two? You mean test returning original when (<large>, None) and (None, <large>)?

> Also about tests, it may be a good idea to create a dedicated demo project just for image tests, so that other build tests don't have to rebuild any of the images.

I agree, but how about not a separate project but simply an image folder and run tests programatically over each of them? With no building involved -- but I'll need a temporary output dir and cleanup afterwards. How would you go about it?

File: lektor/imagetools.py
Comment (nixjdm) [reply lektor/imagetools.py]: Ah, you're right, I did have it. I want to include enum34 so we're actually using enums on 2.7.

My understanding of the codecov report must be off. I though it took into account all testing environments, so it would have hit the exception on the 2.7 tests if enum wasn't available, which it shouldn't have been. So I got confused.

File: lektor/imagetools.py
Comment (nixjdm) [reply lektor/imagetools.py:478]: I didn't think of lektor-thumbnail-generator. In that case we should keep this and keep it as not a private function.

File: lektor/imagetools.py
Comment (nixjdm) [reply lektor/imagetools.py:494]: Makes sense. Thanks.

File: lektor/imagetools.py
Comment (nixjdm) [reply lektor/imagetools.py:576]: I'm having trouble coming up with anything better than "compute_fit_dimensions", which I suppose you're right, doesn't really add much value. It I think of something I'll let you know, but as is is fine for now.

**Comment by nixjdm** [comment]
> You mean to test the warnings and exceptions?

Yes, that's right. Same with `if source_width < width` and `if source_height < height`; I'd like to have tests to make sure we raise exceptions and fallback in expected ways when the thumbnail parameters don't make sense since I expect people to mess up the queries in the templates fairly often.

> I agree, but how about not a separate project but simply an image folder and run tests programatically over each of them? With no building involved -- but I'll need a temporary output dir and cleanup afterwards. How would you go about it?

Do you mean running tests with no ctx / project, or just not performing `build_all`? FWIW I think it's fine to just have another demo project dir with the only interesting content being the images and related template code. I'm not sure how you would test some of these things without a project or ctx, since these images are all subclasses of `Attachment` all the way up to `SourceObject`, which is pretty well tied to needing a ctx and pad.

I think you can call `thumbnail` directly once you have the `Image` though, and not have to build pages, but just make thumbnails, but AFAIK you'd still need a project. You can also use `build` instead of `build_all` and more granular templates instead of one big one to have more specific tests, so you don't have to run through every image every time.

**Comment by xlotlu** [comment]
> Do you mean running tests with no ctx / project, or just not performing build_all?

Yes, no ctx, no project, no "functional" testing, just straightforward unit tests on the imagetools functions.

Then the one thing remaining to need tests with context will be `Image.thumbnail()` -- especially after the compatibility layer, because that's where the warnings will originate from. But I'd do even those with the context mocked if possible, otherwise a minimal project. I wanna stay as unit-ish as possible.

**Comment by xlotlu** [comment]
> Once this is all done, I also request help updating the docs. They'll certainly need it. :)

@nixjdm definitely. I'd have done it already if the docs were part of this repo. Wink, wink. Want me to open a new issue? :)

File: lektor/imagetools.py
Comment (skorokithakis) [reply lektor/imagetools.py:576]: Haha, sorry if I'm making your life harder, guys, let me know if I need to change anything in `thumbnail-generator` and i can do it.

**Comment by xlotlu** [comment]
> 3. optimisation: do find_imagemagick only once. Ideally this should be done once, at startup, outside any process_image request. If too complicated, then at least cache its value.

handled in ~~#559~~ #561.

File: lektor/imagetools.py
Comment (runfalk) [inline lektor/imagetools.py]: Why not fail with a bang here? Isn't the user misusing the API?

**Review by runfalk — CHANGES_REQUESTED** [review]
I think the commit is mostly fine. My major concern is backwards compatibility.

File: tests/test_images.py
Comment (runfalk) [inline tests/test_images.py:118]: We have these kinds of constants in Lektor. They are lowercase, maybe these should be too?

File: lektor/imagetools.py
Comment (xlotlu) [reply lektor/imagetools.py]: true, looks much prettier

File: lektor/imagetools.py
Comment (xlotlu) [reply lektor/imagetools.py]: Yeah, it's useless now since there are only "fit", "crop" and "stretch" modes. But other modes I was considering were "liquid-rescale" and "smart-crop", which would cause accessing `ThumbnailMode.LIQUID_RESCALE` and `ThumbnailMode.SMART_CROP`.

I'll add a bit of docs to that thing.

File: lektor/db.py
Comment (xlotlu) [reply lektor/db.py:783]: yup, the agreement was to build the PR with the final API, and then create one final commit that introduces backward-compatibility (so we can simply revert the commit later on).

File: tests/test_images.py
Comment (xlotlu) [reply tests/test_images.py:118]: or maybe the others should be uppercase? :) I dunno, it's entirely unimportant, especially in that particular place.

File: lektor/db.py
Comment (runfalk) [reply lektor/db.py:783]: I see, that makes sense I guess.

File: lektor/imagetools.py
Comment (nixjdm) [reply lektor/imagetools.py]: Yes, that was it. Rereading the last half of our talking on #540, we were talking about a similar issue but not quite the same thing.

To summarize it all, I think the rule-of-thumb I was going with, is that if you have a template query that can make sense for certain image dimensions, but not others, have a graceful fallback and throw a warning, but don't trip a build failure. For instance, this situation could happen if `upscale=False, mode=crop` and the image is too small to crop for given dimensions. If the query never makes sense though for any image, throw an exception and fail the build. This if block fits this situation, where it never makes sense to crop without given dimensions.

That said, I do not want to break projects that build successfully right now without a transition deprecation period. Currently `.thumbnail(500, crop=True)` builds successfully. I don't want to break that yet. Going with our revert-one-commit-later strategy though, we could throw an exception here, and then dial it back to a deprecation warning in the final commit.

File: tests/test_images.py
Comment (xlotlu) [reply tests/test_images.py:118]: @nixjdm what say you?

Comment (runfalk) [reply tests/test_images.py:118]: I actually was under the impression that constants in PEP8 was snake case. It turns out I was wrong (https://www.python.org/dev/peps/pep-0008/#constants). I guess I'm more in favor of adhering to the existing style (see https://github.com/lektor/lektor/blob/master/lektor/db.py#L793), but in reality I don't really care that much :sweat_smile:.

Comment (nixjdm) [reply tests/test_images.py:118]: Oh yay, a style choice between two recommendations in pep8! @xlotlu keep it the way you've got it. Personally I prefer them being all caps to make them stand out. I've got half a mind to change the others that already exist and make some style rules. Not in this PR though.

**Comment by xlotlu** [comment]
@nixjdm I'm kinda fuzzy and unfocused these days. Help me see if this requires anything else before merging? (besides the final backwards-compatibility commit that is)

**Comment by nixjdm** [comment]
@xlotlu, There's a lot to follow, but I think this is what still should / could be done before the final backwards-compatible-maintaining commit.

1. That bit `if mode != ThumbnailMode.FIT and (width is None or height is None)` should be breaking first, and then the final commit would be non-breaking. It is currently non-breaking, which doesn't fit the plan. I'll re-mark this spot in the diff since the previous conversation went away.
1. (optional) Formatting kwarg; jpg to png and such
1. (optional) Standardizing output filenames. I support this, and agree with your earlier assessment that it isn't really breaking, though it is a bit of file churn for people. Worth it though, IMO.
1. (optional) Additional fluffy modes, e.g. liquid-rescale

Plus, I'm not sure, but it seems like you may want to add a bit to this PR from the further discussion from [this comment](https://github.com/lektor/lektor/pull/497#issuecomment-427786394) on, or if you think all or some of it is better for a third, follow-up PR.

File: lektor/imagetools.py
Comment (nixjdm) [inline lektor/imagetools.py:543]: @xlotlu From previous comment about this, this should be breaking here (as of the latest breaking commit), and in the final backwards-compatible commit it should look like this to be non-breaking.

**Comment by nixjdm** [comment]
@xlotlu That all sounds good to me. And with you adding in that ValueError, this all looks good as the breaking-state. Now please add a single commit to add backwards compatibility, and then this can merge. Post merge, I'll tag that last commit so we don't lose it, for later reversion at an appropriate time.

**Comment by xlotlu** [comment]
I created issues to cover the rest of the changes that were initially considered for this PR: https://github.com/lektor/lektor/issues/607, https://github.com/lektor/lektor/issues/608 and https://github.com/lektor/lektor/issues/609.

**Comment by nixjdm** [comment]
This PR looks good to me, including the final commit, which would be reverted when backwards compatibility is removed in the future.

@runfalk  and @xlotlu Do you have any final thoughts? If not, I'm ready to merge this, and we can shift focus to the video thumbnailing.

@xlotlu thank you for making those new tickets.

**Review by nixjdm — APPROVED** [review]

**Comment by runfalk** [comment]
I just did a quick re-read of the diff. I'm fine with this being merged as it is.

**Review by runfalk — APPROVED** [review]

**Pull request merged**
